indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java

```diff
@@ -698,6 +700,16 @@ private boolean announceTask(
     }
   }
 
+  private boolean cancelWorkerCleanup(String workerHost)
+  {
+    ScheduledFuture previousCleanup = removedWorkerCleanups.remove(workerHost);
+    if (previousCleanup != null) {
+      log.info("Cancelling Worker[%s] scheduled task cleanup", workerHost);
+      previousCleanup.cancel(false);
```

Comment:
The least complex way to do synchronization would be to make the critical sections forced into a single thread. My only concern with that is that we would need to make sure the underlying ZK connection doesn't do weird stuff in the background which renders that thread's state too out of sync.

Also, making sure there won't be anything that hangs up stuff in bad ways. Like making sure cleanup has safe timeouts to try again later in any blocking section so that new workers can be handled. I'd hate to have a cleanup thread block the ability to launch new tasks because the runner thinks the cluster is at capacity.